hook-worker/src/worker.rs

```diff
@@ -163,70 +178,104 @@ impl<'p> WebhookWorker<'p> {
                 .set(1f64 - semaphore.available_permits() as f64 / self.max_concurrent_jobs as f64);
         };
 
+        let dequeue_batch_size_histogram = metrics::histogram!("webhook_dequeue_batch_size");
+
         if transactional {
             loop {
                 report_semaphore_utilization();
-                let webhook_job = self.wait_for_job_tx().await;
-                spawn_webhook_job_processing_task(
-                    self.client.clone(),
-                    semaphore.clone(),
-                    self.retry_policy.clone(),
-                    webhook_job,
-                )
-                .await;
+                // TODO: We could grab semaphore permits here using something like:
+                //   `min(semaphore.available_permits(), dequeue_batch_size)`
+                // And then dequeue only up to that many jobs. We'd then need to hand back the
+                // difference in permits based on how many jobs were dequeued.
+                let mut batch = self.wait_for_jobs_tx().await;
+                dequeue_batch_size_histogram.record(batch.jobs.len() as f64);
+
+                // Get enough permits for the jobs before spawning a task.
+                let permits = semaphore
+                    .clone()
+                    .acquire_many_owned(batch.jobs.len() as u32)
+                    .await
+                    .expect("semaphore has been closed");
+
+                let client = self.client.clone();
+                let retry_policy = self.retry_policy.clone();
+
+                tokio::spawn(async move {
+                    let mut futures = Vec::new();
+
+                    // We have to `take` the Vec of jobs from the batch to avoid a borrow checker
+                    // error below when we commit.
+                    for job in std::mem::take(&mut batch.jobs) {
```

Comment:
This makes me a bit nervous: The borrow checker is there to help us, not sure I want to _avoid_ it.

Reply:
The borrow checker is still in play, maybe my comment is just bad. I can't find a way to mutate each job in the Vec (via `complete()` etc) *and* still be allowed to commit at the end.